File: blackjack.cpp
// blackjack.cpp
#include <iostream>
#include <vector>
#include <string>
#include <random>
#include <algorithm>
#include <map>
#include <fstream>
#include <cctype>
#include <filesystem>

using namespace std;
namespace fs = std::filesystem;

const string RESET = "\033[0m";
const string RED = "\033[91m";
const string GREEN = "\033[92m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string CYAN = "\033[96m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

vector<string> SUITS = {"♠", "♥", "♦", "♣"};
vector<string> RANKS = {"2","3","4","5","6","7","8","9","10","J","Q","K","A"};
map<string, int> VALUES = {{"2",2},{"3",3},{"4",4},{"5",5},{"6",6},{"7",7},{"8",8},{"9",9},{"10",10},{"J",10},{"Q",10},{"K",10},{"A",11}};

struct Card {
    string suit;
    string rank;
    int value;
    string color;
    Card(string s, string r) : suit(s), rank(r), value(VALUES[r]) {
        color = (s == "♥" || s == "♦") ? RED : RESET;
    }
    string str() const {
        return colorize(rank + suit, color);
    }
};

class Deck {
public:
    vector<Card> cards;
    Deck() {
        for (auto& s : SUITS)
            for (auto& r : RANKS)
                cards.push_back(Card(s, r));
        shuffle();
    }
    void shuffle() {
        random_device rd;
        mt19937 g(rd());
        ::shuffle(cards.begin(), cards.end(), g);
    }
    Card draw() {
        if (cards.empty()) return Card("", "");
        Card c = cards.back();
        cards.pop_back();
        return c;
    }
};

class Hand {
public:
    vector<Card> cards;
    int bet;
    bool done;
    Hand() : bet(0), done(false) {}
    void add(Card c) { cards.push_back(c); }
    int value() const {
        int total = 0;
        int aces = 0;
        for (auto& c : cards) {
            total += c.value;
            if (c.rank == "A") aces++;
        }
        while (total > 21 && aces > 0) {
            total -= 10;
            aces--;
        }
        return total;
    }
    bool isBlackjack() const {
        return cards.size() == 2 && value() == 21;
    }
    bool isBust() const { return value() > 21; }
    bool canSplit() const {
        return cards.size() == 2 && cards[0].rank == cards[1].rank;
    }
    Hand split() {
        Hand h;
        h.cards.push_back(cards.back());
        cards.pop_back();
        return h;
    }
    string str() const {
        string s;
        for (auto& c : cards) s += c.str() + " ";
        return s;
    }
};

class Blackjack {
public:
    int balance;
    int bet;
    Deck deck;
    vector<Hand> player_hands;
    Hand dealer_hand;
    bool insurance;
    int insurance_paid;
    map<string, int> stats;
    string stats_file;

    Blackjack(int bal) : balance(bal), bet(0), insurance(false), insurance_paid(0) {
        stats_file = string(getenv("HOME")) + "/.blackjack_stats.json";
        loadStats();
    }

    void loadStats() {
        ifstream f(stats_file);
        if (!f) return;
        string line;
        while (getline(f, line)) {
            size_t p = line.find(':');
            if (p != string::npos) {
                string key = line.substr(0, p);
                int val = stoi(line.substr(p+1));
                stats[key] = val;
            }
        }
        if (stats.find("wins") == stats.end()) {
            stats["wins"] = 0; stats["losses"] = 0; stats["pushes"] = 0; stats["blackjacks"] = 0;
        }
    }

    void saveStats() {
        ofstream f(stats_file);
        if (f) {
            f << "wins:" << stats["wins"] << "\n";
            f << "losses:" << stats["losses"] << "\n";
            f << "pushes:" << stats["pushes"] << "\n";
            f << "blackjacks:" << stats["blackjacks"] << "\n";
        }
    }

    void displayStats() {
        cout << colorize("Статистика:", BOLD) << endl;
        cout << "  Побед: " << stats["wins"] << endl;
        cout << "  Поражений: " << stats["losses"] << endl;
        cout << "  Ничьих: " << stats["pushes"] << endl;
        cout << "  Блэкджеков: " << stats["blackjacks"] << endl;
        cout << "  Баланс: " << balance << endl;
    }

    void dealInitial() {
        deck = Deck();
        deck.shuffle();
        player_hands.clear();
        player_hands.push_back(Hand());
        dealer_hand = Hand();
        insurance = false;
        insurance_paid = 0;
        player_hands[0].add(deck.draw());
        dealer_hand.add(deck.draw());
        player_hands[0].add(deck.draw());
        dealer_hand.add(deck.draw());
    }

    void showHands(bool hideDealer = true) {
        cout << "\nДилер:" << endl;
        if (hideDealer) {
            cout << "  " << dealer_hand.cards[0].str() << "  [скрыто]" << endl;
        } else {
            cout << "  " << dealer_hand.str() << " (очки: " << dealer_hand.value() << ")" << endl;
        }
        cout << "\nВаши руки:" << endl;
        for (size_t i=0; i<player_hands.size(); ++i) {
            cout << "  Рука " << i+1 << ": " << player_hands[i].str() << " (очки: " << player_hands[i].value() << ")" << endl;
        }
    }

    void getBet() {
        while (true) {
            cout << "Ваша ставка (баланс: " << balance << "): ";
            string inp;
            cin >> inp;
            try {
                int b = stoi(inp);
                if (b > 0 && b <= balance) {
                    bet = b;
                    return;
                } else {
                    cout << colorize("Неверная ставка.", RED) << endl;
                }
            } catch (...) {
                cout << colorize("Введите число.", RED) << endl;
            }
        }
    }

    void playHand(int idx) {
        Hand& hand = player_hands[idx];
        while (!hand.done && !hand.isBust() && hand.value() < 21) {
            showHands();
            cout << "\nРука " << idx+1 << ": " << hand.str() << " (очки: " << hand.value() << ")" << endl;
            cout << "Действие (h=hit, s=stand, d=double, i=insurance, sp=split, q=quit): ";
            string action;
            cin >> action;
            if (action == "q") {
                cout << colorize("Выход.", YELLOW) << endl;
                saveStats();
                exit(0);
            } else if (action == "h") {
                hand.add(deck.draw());
                if (hand.isBust()) {
                    cout << colorize("Перебор!", RED) << endl;
                    hand.done = true;
                }
            } else if (action == "s") {
                hand.done = true;
            } else if (action == "d") {
                if (hand.cards.size() == 2 && balance >= bet) {
                    bet *= 2;
                    hand.add(deck.draw());
                    hand.done = true;
                    if (hand.isBust()) cout << colorize("Перебор!", RED) << endl;
                } else {
                    cout << colorize("Удвоение недоступно.", YELLOW) << endl;
                }
            } else if (action == "i") {
                if (dealer_hand.cards[0].rank == "A" && !insurance) {
                    insurance = true;
                    insurance_paid = bet / 2;
                    balance -= insurance_paid;
                    cout << colorize("Страховка активирована.", CYAN) << endl;
                } else {
                    cout << colorize("Страховка недоступна.", YELLOW) << endl;
                }
            } else if (action == "sp") {
                if (hand.canSplit()) {
                    Hand new_hand = hand.split();
                    player_hands.push_back(new_hand);
                    hand.add(deck.draw());
                    new_hand.add(deck.draw());
                    cout << colorize("Руки разделены.", CYAN) << endl;
                    if (balance < bet) {
                        cout << colorize("Недостаточно средств для сплита.", RED) << endl;
                        // откат
                        player_hands.pop_back();
                    } else {
                        balance -= bet;
                        bet *= 2;
                    }
                } else {
                    cout << colorize("Сплит недоступен.", YELLOW) << endl;
                }
            } else {
                cout << colorize("Неизвестная команда.", RED) << endl;
            }
        }
    }

    void dealerPlay() {
        while (dealer_hand.value() < 17) {
            dealer_hand.add(deck.draw());
        }
    }

    void resolve() {
        showHands(false);
        int dealer_val = dealer_hand.value();
        for (auto& hand : player_hands) {
            int player_val = hand.value();
            if (hand.isBust()) {
                balance -= bet;
                stats["losses"]++;
                cout << colorize("Поражение (перебор): рука проиграла " + to_string(bet), RED) << endl;
            } else if (hand.isBlackjack() && !dealer_hand.isBlackjack()) {
                int win = bet * 1.5;
                balance += win;
                stats["blackjacks"]++;
                stats["wins"]++;
                cout << colorize("Блэкджек! Выигрыш " + to_string(win), GREEN) << endl;
            } else if (dealer_val > 21) {
                balance += bet;
                stats["wins"]++;
                cout << colorize("Дилер перебрал, выигрыш " + to_string(bet), GREEN) << endl;
            } else if (player_val > dealer_val) {
                balance += bet;
                stats["wins"]++;
                cout << colorize("Выигрыш " + to_string(bet), GREEN) << endl;
            } else if (player_val == dealer_val) {
                stats["pushes"]++;
                cout << colorize("Ничья, ставка возвращена.", YELLOW) << endl;
            } else {
                balance -= bet;
                stats["losses"]++;
                cout << colorize("Поражение, потеряно " + to_string(bet), RED) << endl;
            }
        }
        if (insurance) {
            if (dealer_hand.isBlackjack()) {
                balance += insurance_paid * 2;
                cout << colorize("Страховка выиграла!", CYAN) << endl;
            } else {
                cout << colorize("Страховка проиграла.", YELLOW) << endl;
            }
        }
        saveStats();
    }

    void play() {
        cout << colorize("🃏 Добро пожаловать в Блэкджек!", BOLD) << endl;
        cout << "Ваш баланс: " << balance << endl;
        while (balance > 0) {
            getBet();
            dealInitial();
            // Блэкджек игрока
            if (player_hands[0].isBlackjack()) {
                showHands();
                if (dealer_hand.isBlackjack()) {
                    cout << colorize("Ничья – у обоих блэкджек!", YELLOW) << endl;
                    stats["pushes"]++;
                    saveStats();
                    continue;
                } else {
                    int win = bet * 1.5;
                    balance += win;
                    stats["blackjacks"]++;
                    stats["wins"]++;
                    cout << colorize("Блэкджек! Выигрыш " + to_string(win), GREEN) << endl;
                    saveStats();
                    continue;
                }
            }
            // Страховка
            if (dealer_hand.cards[0].rank == "A") {
                cout << "У дилера туз. Хотите страховку? (y/n): ";
                string ans;
                cin >> ans;
                if (ans == "y") {
                    insurance = true;
                    insurance_paid = bet / 2;
                    balance -= insurance_paid;
                    cout << colorize("Страховка активирована.", CYAN) << endl;
                }
            }
            // Ход игрока
            for (size_t i=0; i<player_hands.size(); ++i) {
                playHand(i);
            }
            // Ход дилера
            dealerPlay();
            resolve();
            cout << colorize("Баланс: " + to_string(balance), BOLD) << endl;
            if (balance <= 0) {
                cout << colorize("Вы проиграли все деньги!", RED) << endl;
                break;
            }
            cout << "Продолжить игру? (y/n): ";
            string cont;
            cin >> cont;
            if (cont != "y") break;
        }
        displayStats();
        saveStats();
    }
};

int main(int argc, char* argv[]) {
    int balance = 1000;
    bool reset = false;
    for (int i=1; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "-r" || arg == "--reset") reset = true;
        else if ((arg == "-s" || arg == "--start") && i+1 < argc) {
            balance = stoi(argv[++i]);
        } else if (arg == "-h" || arg == "--help") {
            cout << "Usage: blackjack [-s start_balance] [-r]" << endl;
            return 0;
        }
    }
    if (reset) {
        string home = getenv("HOME");
        string f = home + "/.blackjack_stats.json";
        if (fs::exists(f)) fs::remove(f);
        cout << "Статистика сброшена." << endl;
        return 0;
    }
    Blackjack game(balance);
    game.play();
    return 0;
}

File: blackjack.java
// blackjack.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class blackjack {
    private static final String RESET = "\u001B[0m";
    private static final String RED = "\u001B[91m";
    private static final String GREEN = "\u001B[92m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String CYAN = "\u001B[96m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static final String[] SUITS = {"♠","♥","♦","♣"};
    private static final String[] RANKS = {"2","3","4","5","6","7","8","9","10","J","Q","K","A"};
    private static final Map<String, Integer> VALUES = new HashMap<>();
    static {
        VALUES.put("2",2); VALUES.put("3",3); VALUES.put("4",4); VALUES.put("5",5);
        VALUES.put("6",6); VALUES.put("7",7); VALUES.put("8",8); VALUES.put("9",9);
        VALUES.put("10",10); VALUES.put("J",10); VALUES.put("Q",10); VALUES.put("K",10); VALUES.put("A",11);
    }

    static class Card {
        String suit, rank;
        int value;
        Card(String s, String r) { suit=s; rank=r; value=VALUES.get(r); }
        @Override
        public String toString() {
            String col = (suit.equals("♥") || suit.equals("♦")) ? RED : RESET;
            return colorize(rank + suit, col);
        }
    }

    static class Deck {
        List<Card> cards;
        Deck() {
            cards = new ArrayList<>();
            for (String s : SUITS) for (String r : RANKS) cards.add(new Card(s, r));
            shuffle();
        }
        void shuffle() {
            Collections.shuffle(cards);
        }
        Card draw() {
            if (cards.isEmpty()) return null;
            return cards.remove(cards.size()-1);
        }
    }

    static class Hand {
        List<Card> cards = new ArrayList<>();
        int bet;
        boolean done;
        int value() {
            int total = cards.stream().mapToInt(c -> c.value).sum();
            int aces = (int) cards.stream().filter(c -> c.rank.equals("A")).count();
            while (total > 21 && aces > 0) { total -= 10; aces--; }
            return total;
        }
        boolean isBlackjack() { return cards.size() == 2 && value() == 21; }
        boolean isBust() { return value() > 21; }
        boolean canSplit() { return cards.size() == 2 && cards.get(0).rank.equals(cards.get(1).rank); }
        Hand split() {
            Hand h = new Hand();
            h.cards.add(cards.remove(cards.size()-1));
            return h;
        }
        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            for (Card c : cards) sb.append(c.toString()).append(" ");
            return sb.toString().trim();
        }
    }

    static class Stats {
        int wins, losses, pushes, blackjacks;
    }

    static class Blackjack {
        int balance, bet;
        Deck deck;
        List<Hand> playerHands;
        Hand dealerHand;
        boolean insurance;
        int insurancePaid;
        Stats stats = new Stats();
        String statsFile;

        Blackjack(int bal) {
            balance = bal;
            statsFile = System.getProperty("user.home") + "/.blackjack_stats.json";
            loadStats();
        }

        void loadStats() {
            try {
                String json = new String(Files.readAllBytes(Paths.get(statsFile)));
                // упрощённый парсинг (без библиотек)
                stats.wins = extractInt(json, "wins");
                stats.losses = extractInt(json, "losses");
                stats.pushes = extractInt(json, "pushes");
                stats.blackjacks = extractInt(json, "blackjacks");
            } catch (Exception e) {}
        }

        int extractInt(String json, String key) {
            int idx = json.indexOf("\"" + key + "\":");
            if (idx == -1) return 0;
            int start = idx + key.length() + 3;
            int end = json.indexOf(",", start);
            if (end == -1) end = json.indexOf("}", start);
            try { return Integer.parseInt(json.substring(start, end).trim()); } catch (Exception e) { return 0; }
        }

        void saveStats() {
            try {
                String json = "{\"wins\":"+stats.wins+",\"losses\":"+stats.losses+
                              ",\"pushes\":"+stats.pushes+",\"blackjacks\":"+stats.blackjacks+"}";
                Files.write(Paths.get(statsFile), json.getBytes());
            } catch (Exception e) {}
        }

        void displayStats() {
            System.out.println(colorize("Статистика:", BOLD));
            System.out.println("  Побед: " + stats.wins);
            System.out.println("  Поражений: " + stats.losses);
            System.out.println("  Ничьих: " + stats.pushes);
            System.out.println("  Блэкджеков: " + stats.blackjacks);
            System.out.println("  Баланс: " + balance);
        }

        void dealInitial() {
            deck = new Deck();
            deck.shuffle();
            playerHands = new ArrayList<>();
            playerHands.add(new Hand());
            dealerHand = new Hand();
            insurance = false;
            insurancePaid = 0;
            playerHands.get(0).cards.add(deck.draw());
            dealerHand.cards.add(deck.draw());
            playerHands.get(0).cards.add(deck.draw());
            dealerHand.cards.add(deck.draw());
        }

        void showHands(boolean hideDealer) {
            System.out.println("\nДилер:");
            if (hideDealer) {
                System.out.println("  " + dealerHand.cards.get(0).toString() + "  [скрыто]");
            } else {
                System.out.println("  " + dealerHand.toString() + " (очки: " + dealerHand.value() + ")");
            }
            System.out.println("\nВаши руки:");
            for (int i=0; i<playerHands.size(); i++) {
                System.out.println("  Рука " + (i+1) + ": " + playerHands.get(i).toString() + " (очки: " + playerHands.get(i).value() + ")");
            }
        }

        void getBet() {
            Scanner sc = new Scanner(System.in);
            while (true) {
                System.out.print("Ваша ставка (баланс: " + balance + "): ");
                String inp = sc.nextLine();
                try {
                    int b = Integer.parseInt(inp);
                    if (b > 0 && b <= balance) { bet = b; return; }
                } catch (Exception e) {}
                System.out.println(colorize("Неверная ставка.", RED));
            }
        }

        void playHand(int idx, Scanner sc) {
            Hand hand = playerHands.get(idx);
            while (!hand.done && !hand.isBust() && hand.value() < 21) {
                showHands(true);
                System.out.println("\nРука " + (idx+1) + ": " + hand.toString() + " (очки: " + hand.value() + ")");
                System.out.print("Действие (h=hit, s=stand, d=double, i=insurance, sp=split, q=quit): ");
                String action = sc.nextLine().trim().toLowerCase();
                switch (action) {
                    case "q": System.out.println(colorize("Выход.", YELLOW)); saveStats(); System.exit(0); break;
                    case "h":
                        hand.cards.add(deck.draw());
                        if (hand.isBust()) { System.out.println(colorize("Перебор!", RED)); hand.done = true; }
                        break;
                    case "s": hand.done = true; break;
                    case "d":
                        if (hand.cards.size() == 2 && balance >= bet) {
                            bet *= 2;
                            hand.cards.add(deck.draw());
                            hand.done = true;
                            if (hand.isBust()) System.out.println(colorize("Перебор!", RED));
                        } else System.out.println(colorize("Удвоение недоступно.", YELLOW));
                        break;
                    case "i":
                        if (dealerHand.cards.get(0).rank.equals("A") && !insurance) {
                            insurance = true;
                            insurancePaid = bet / 2;
                            balance -= insurancePaid;
                            System.out.println(colorize("Страховка активирована.", CYAN));
                        } else System.out.println(colorize("Страховка недоступна.", YELLOW));
                        break;
                    case "sp":
                        if (hand.canSplit()) {
                            Hand newHand = hand.split();
                            playerHands.add(newHand);
                            hand.cards.add(deck.draw());
                            newHand.cards.add(deck.draw());
                            System.out.println(colorize("Руки разделены.", CYAN));
                            if (balance < bet) {
                                System.out.println(colorize("Недостаточно средств для сплита.", RED));
                                playerHands.remove(playerHands.size()-1);
                            } else {
                                balance -= bet;
                                bet *= 2;
                            }
                        } else System.out.println(colorize("Сплит недоступен.", YELLOW));
                        break;
                    default: System.out.println(colorize("Неизвестная команда.", RED));
                }
            }
        }

        void dealerPlay() {
            while (dealerHand.value() < 17) {
                dealerHand.cards.add(deck.draw());
            }
        }

        void resolve() {
            showHands(false);
            int dealerVal = dealerHand.value();
            for (Hand hand : playerHands) {
                int playerVal = hand.value();
                if (hand.isBust()) {
                    balance -= bet;
                    stats.losses++;
                    System.out.println(colorize("Поражение (перебор): рука проиграла " + bet, RED));
                } else if (hand.isBlackjack() && !dealerHand.isBlackjack()) {
                    int win = (int)(bet * 1.5);
                    balance += win;
                    stats.blackjacks++;
                    stats.wins++;
                    System.out.println(colorize("Блэкджек! Выигрыш " + win, GREEN));
                } else if (dealerVal > 21) {
                    balance += bet;
                    stats.wins++;
                    System.out.println(colorize("Дилер перебрал, выигрыш " + bet, GREEN));
                } else if (playerVal > dealerVal) {
                    balance += bet;
                    stats.wins++;
                    System.out.println(colorize("Выигрыш " + bet, GREEN));
                } else if (playerVal == dealerVal) {
                    stats.pushes++;
                    System.out.println(colorize("Ничья, ставка возвращена.", YELLOW));
                } else {
                    balance -= bet;
                    stats.losses++;
                    System.out.println(colorize("Поражение, потеряно " + bet, RED));
                }
            }
            if (insurance) {
                if (dealerHand.isBlackjack()) {
                    balance += insurancePaid * 2;
                    System.out.println(colorize("Страховка выиграла!", CYAN));
                } else System.out.println(colorize("Страховка проиграла.", YELLOW));
            }
            saveStats();
        }

        void play() {
            Scanner sc = new Scanner(System.in);
            System.out.println(colorize("🃏 Добро пожаловать в Блэкджек!", BOLD));
            System.out.println("Ваш баланс: " + balance);
            while (balance > 0) {
                getBet();
                dealInitial();
                if (playerHands.get(0).isBlackjack()) {
                    showHands(true);
                    if (dealerHand.isBlackjack()) {
                        System.out.println(colorize("Ничья – у обоих блэкджек!", YELLOW));
                        stats.pushes++;
                        saveStats();
                        continue;
                    } else {
                        int win = (int)(bet * 1.5);
                        balance += win;
                        stats.blackjacks++;
                        stats.wins++;
                        System.out.println(colorize("Блэкджек! Выигрыш " + win, GREEN));
                        saveStats();
                        continue;
                    }
                }
                if (dealerHand.cards.get(0).rank.equals("A")) {
                    System.out.print("У дилера туз. Хотите страховку? (y/n): ");
                    String ans = sc.nextLine().trim().toLowerCase();
                    if (ans.equals("y")) {
                        insurance = true;
                        insurancePaid = bet / 2;
                        balance -= insurancePaid;
                        System.out.println(colorize("Страховка активирована.", CYAN));
                    }
                }
                for (int i=0; i<playerHands.size(); i++) {
                    playHand(i, sc);
                }
                dealerPlay();
                resolve();
                System.out.println(colorize("Баланс: " + balance, BOLD));
                if (balance <= 0) {
                    System.out.println(colorize("Вы проиграли все деньги!", RED));
                    break;
                }
                System.out.print("Продолжить игру? (y/n): ");
                String cont = sc.nextLine().trim().toLowerCase();
                if (!cont.equals("y")) break;
            }
            displayStats();
            saveStats();
            sc.close();
        }
    }

    public static void main(String[] args) {
        int balance = 1000;
        boolean reset = false;
        for (int i=0; i<args.length; i++) {
            if (args[i].equals("-r") || args[i].equals("--reset")) reset = true;
            else if ((args[i].equals("-s") || args[i].equals("--start")) && i+1 < args.length) {
                balance = Integer.parseInt(args[++i]);
            } else if (args[i].equals("-h") || args[i].equals("--help")) {
                System.out.println("Usage: java blackjack [-s start_balance] [-r]");
                return;
            }
        }
        if (reset) {
            String f = System.getProperty("user.home") + "/.blackjack_stats.json";
            try { Files.deleteIfExists(Paths.get(f)); } catch (Exception e) {}
            System.out.println("Статистика сброшена.");
            return;
        }
        Blackjack game = new Blackjack(balance);
        game.play();
    }
}
